File: Data-Structure/_All-In-One-LeetCode/00001_twoSum (E)/twoSum.js
const twoSum = (arr, target) => {
    if(arr.length < 2) return [];
    let res = [];
    let map = new Map();
    let N = arr.length;
    for (let i = 0; i < N; i++) {
        let key = arr[i];
        if (map.has(target - key)) {
            return [map.get(target - key) ,  i];
        }
        map.set(key, i);
    }
    return res;
}

File: Data-Structure/_All-In-One-LeetCode/00001_twoSum (E)/two_sum.py
from ast import List

"""
enumerate is useful for obtaining an indexed list:
    (0, seq[0]), (1, seq[1]), (2, seq[2]), ...
"""


class Solution:
    def two_sum(self, nums: List[int], target: int) -> List[int]:
        hashmap = {}  # k = val, v = index

        for i, n in enumerate(nums):
            diff = target - n
            if diff in hashmap:
                return [hashmap[diff], i]
            hashmap[n] = i

        return
